appinventor/components/src/com/google/appinventor/components/runtime/util/EclairUtil.java

```diff
@@ -46,44 +76,177 @@ public static void overridePendingTransitions(Activity activity, int enterAnim,
+            sep = ", ";
+          }
+
+          return accessStringBuilder.toString();
+        }
+
+        /**
+         * Auxiliary function to extract all the required Android
+         * permissions from a single PermissionRequest.
+         *
+         * @return Array of permissions
+         */
+        private String[] getPermissionsForRequest(final PermissionRequest request) {
+          // Get requested resources
+          String[] resources = request.getResources();
+
+          // Maintain list of all permissions needed
+          List<String> permissions = new ArrayList<String>();
+
+          for (String res : resources) {
+            // Mapping exists (note that PERMISSION_MAP only contains a subset of RESOURCE_MAP entries,
+            // so all other cases are not covered by the mapping)
+            if (RESOURCE_MAP.containsKey(res)) {
+              // Map res -> RESOURCE_MAP name
+              final String resourceName = RESOURCE_MAP.get(res);
+
+              // Retrieve permissions (if they exist), and convert to list
+              final List<String> newPermissions = Arrays.asList(PERMISSION_MAP.getOrDefault(resourceName, new String[] {}));
+
+              // Concatenate lists together
+              permissions.addAll(newPermissions);
+            }
+          }
+
+          // Return concatenated permissions as an array of strings
+          return permissions.toArray(new String[permissions.size()]);
+        }
+
+        /**
+         * Creates a permission alert dialog that perform the corresponding listener actions
+         * on accept or reject.
+         * @param origin  Origin of the permission request (URL string)
+         * @param access  String indicating what the permission wants to access
+         * @param acceptListener OnClickListener for accepting
+         * @param refuseListener OnClickListener for refusing
+         */
+        private void createPermissionAlertDialog(String origin, String access, DialogInterface.OnClickListener acceptListener,
+          DialogInterface.OnClickListener refuseListener) {
+            AlertDialog alertDialog = new AlertDialog.Builder(activity).create();
+
+            alertDialog.setTitle("Permission Request");
+
+            if (origin.equals("file://"))
+              origin = "This Application";
+
+            alertDialog.setMessage(origin + " would like to access your " + access + ".");
+            
+            alertDialog.setButton(AlertDialog.BUTTON_POSITIVE, "Allow", acceptListener);
+            alertDialog.setButton(AlertDialog.BUTTON_NEGATIVE, "Refuse", refuseListener);
+            
+            alertDialog.show();
+        }
+
+        /**
+         * Auxiliary function to request Android permissions from the specified array
+         * of strings.
+         *
+         * @param permissions Array of permissions to request
+         */
+        private void resolveAndroidPermissions(String[] permissions) {
```

Comment:
Things have became quite trickier and messier in new APIs.. When testing on API26, the WebViewer will not prompt for Android permissions at all when using WebView; The WebView permission request & the Android one are two completely separate permission requests, and by default, most permissions are disabled (e.g. Location is disabled by default)

As such, it was necessary to also prompt for manual permission requesting

Reply:
This is what I was expecting. Apps request permission for these permissions, and then access can be granted on a per website basis by the user through the prompts. Otherwise, once the app is granted location permission every website would be able to read it.

Reply:
Yeah, exactly. I agree with the current pattern used in `EclairUtil` prior to the PR, but was surprised to see that the functionality no longer works on higher APIs; What do you think of the current workaround?

Reply:
My guess is this is affected by whether Android is using the built-in Android system webview or the Google Chrome webview. It likely will also be affected by other platform, like Amazon Fire, which has an entirely different webview implementation. I think we may just need to do more experimentation here.